packages/core-data/src/dynamic-selectors.ts

```diff
@@ -0,0 +1,369 @@
+/**
+ * Internal dependencies
+ */
+import type { GetRecordsHttpQuery, State } from './selectors';
+import type * as ET from './entity-types';
+
+type Nullable< T > = T | null;
+
+/**
+ * Get post type object by name.
+ *
+ * @param state Data state.
+ * @param name  Post type name.
+ * @param query Optional query. If requesting specific
+ */
+export declare function getPostType(
+	state: State,
+	name: string,
+	query?: GetRecordsHttpQuery
+): ET.PostType | undefined;
```

Comment:
I'm curious about this style of `declare function…`, has this been used elsewhere to declare types for core data?

It concerns me because `declare …` is similar to a type assertion, it's basically introducing a typed value into the type system that it's unaware of. It manifests something that isn't there. That's dangerous:

```ts
declare function foo(): void;
foo(); // no type error, runtime error.
```

Can these be function interfaces (or type aliases) instead?

```suggestion
/**
 * Get post type object by name.
 *
 * @param state Data state.
 * @param name  Post type name.
 * @param query Optional query. If requesting specific
 */
export declare function getPostType(
	state: State,
	name: string,
	query?: GetRecordsHttpQuery
): ET.PostType | undefined;
```

I worry otherwise that these functions (that don't actually exist) will be available to import from this module.

I haven't tested to confirm this, but it's my understanding of `declare`.

Reply:
Using interfaces and type aliases will not work in this case because you can't import a type and expect it to be available as value.

I looked precisely into this concern when I started on this, and you are right that `declare` is an assertion, and that is exactly what needs to be done here. We are asserting that this has been defined dynamically elsewhere.

An improvement we can have here is to probably declare a namespace and move these declarations into it to avoid direct imports. That way, it these functions won't be available for direct import.


```ts
export namespace DynamicSelectors {
	/**
	 * Get post type object by name.
	 *
	 * @param state Data state.
	 * @param name  Post type name.
	 * @param query Optional query. If requesting specific
	 */
	export declare function getPostType(
		state: State,
		name: string,
		query?: GetRecordsHttpQuery
	): ET.PostType | undefined;

	// Other stuff
}
```